### PR TITLE
Refactorings and updates

### PR DESCRIPTION
**Pull request by philiptaron**
Hello there! I thought I might provide a few updates and refactorings to this useful set of scripts.

1. Update to `nixfmt-rfc-style` ([RFC 166](https://github.com/NixOS/rfcs/blob/master/rfcs/0166-nix-formatting.md)).
2. Update flake.lock and change the input to nixos-unstable to match default.nix.
3. Create a package.nix and use [`callPackage` style](https://nix.dev/tutorials/callpackage) to inject dependencies.
4. Unify the default revision being used to pull from `flake.lock` (without taking a dependency on flakes themselves.)
5. Move the app definitions into the package, rather than having them be free-floating in the flake. Maybe a new pattern?
6. <del>Use https://github.com/nix-systems/nix-systems to inject the list of systems, so they're externally override-able, rather than encoding that as a list in the flake.</del> Reverted per feedback.
7. Use `nixpkgs.legacyPackages.${system}` in the flake instead of re-importing `nixpkgs` (c.f. https://zimbatm.com/notes/1000-instances-of-nixpkgs)
8. Wired up the formatter value in outputs so that `nix fmt .` works.
9. Added `result*` to `.gitignore` so that `nix build` won't leave untracked files that might accidentally added.

### CODE REVIEW

**Comment by philiptaron** [comment]
@n8henrie @timokau as the relevant folks

**Comment by n8henrie** [comment]
Thanks for your updates!

- `Update to nixfmt-rfc-style`
    - I agree
- Update flake.lock and inputs to reflect NixOS 23.11 being out.
    - I pinned in my original "use flakes" PR, but I think following `nixpkgs-unstable` seems to be the common practice and seems quite reasonable since we're publishing a `flake.lock` anyways. Would prevent the need to keep this up to date.
- Create a package.nix and use callPackage style to inject dependencies.
    - `callPackage` provides a useful means for overriding inputs but -- for most of my simple use-cases that never end up benefiting from these -- also feels like a premature optimization. Since the work is already done I don't see why not, and I'm open to learning if there are other benefits.
- Unify the default revision being used to pull from flake.lock (without taking a dependency on flakes themselves.)
    - Makes sense -- though couldn't one just remove all that and just expect `pkgs` to be passed in? I've never used / learned non-flakes nix, but I thought `pkgs` was implicitly passed in (and would just default to the system `<nixpkgs>` or something). 
- Move the app definitions into the package, rather than having them be free-floating in the flake.
    - Just curious about the reasoning here, I haven't seen this pattern before and I get 0 matches for `rg -F passthru.apps` in all of nixpkgs. Does this perhaps allow running the default app in non-flake systems?
- Use https://github.com/nix-systems/nix-systems to inject the list of systems, so they're externally override-able, rather than encoding that as a list in the flake.
    - I don't see the benefit here, kind of seems like another premature optimization and another dependency to track and update to boot (it looks like all commits on this flake were made between April and July 2023). I would recommend just adding `x86_64-darwin` (assuming we can find a tester) and that would provide all the same outputs by default, and handling requests to provide additional architectures on an as-needed basis. Outside of the big 4, what other systems do you think people are using to bisect nixpkgs? I see the logic in letting the users 
- Use nixpkgs.legacyPackages.${system} in the flake instead of re-importing nixpkgs (c.f. https://zimbatm.com/notes/1000-instances-of-nixpkgs)
    - I'm happy to be corrected here, but I don't think that article applies when importing nixpkgs *without specifying an overlay*. I think `legacyPackages` [just imports nixpkgs](https://github.com/NixOS/nixpkgs/blob/ad9440ea07443100ac61d3a1cde68de451afbaae/flake.nix#L85) just like we're doing. If you know differently please let me know.

Overall I think this looks fine. I think a few of the changes slightly increase complexity without much (current) benefit as far as I can tell, though may have future benefits. I definitely agree with updating the inputs and formatting.

As it's semi-relevant to keeping things simple, for my own projects I've also settled on a somewhat simpler `systemClosure` function than I used in this project (source: https://www.reddit.com/r/NixOS/comments/13nphq2/comment/jl30akk):

```nix
systems = ["aarch64-darwin" "x86_64-linux" "aarch64-linux"];
eachSystem = with nixpkgs.lib;
  f:
    foldAttrs mergeAttrs {}
    (map (s: mapAttrs (_: v: {${s} = v;}) (f s)) systems);
```

and I think it would also be reasonable to use the even simpler version 

```nix
let 
systems =  [
    "x86_64-linux"
    "aarch64-linux"
  ];
forAllSystems = function:
  nixpkgs.lib.genAttrs systems (system:
    function (import nixpkgs {
      inherit system;
    }));
in
```

**Comment by philiptaron** [comment]
> Unstable or a release?

I also like unstable. I followed the existing flake for hints as to which version is preferred here, and I'll update it to that branch by default.

> Couldn't one just remove all that and just expect pkgs to be passed in?

My understanding is no; any sort of "defaulting" mechanism needs to be built oneself, which is why you see the diversity of ways of doing that across non-flakes Nix.

> callPackage

I really want this one 🙂. 

1. Because the packaging code is not in `default.nix`, it's a clear separation of responsibilities.
2. Because the derivation isn't tied to the `python3` symbol, it lets other `python` packages (say: `pypy`, `python312`, etc.) splice in their own set of packages.
3. Because the derivation is in its own file, I can import it in my own overlay and fit it in.
4. It's the style that would be insisted on inside `nixpkgs`, so those reading nixpkgs for examples will be familiar with this [callPackage-style](https://nix.dev/tutorials/callpackage).

> passthru.apps

There are two reasons I did this:

1. Attach the apps to the package, so that the names of scripts being pulled out were inside of the derivation that contained them. This was the biggest one for me, since I found the action at a distance of the derivation and the thing pulling names out of the derivation unseemly.
5. Allow access to the apps outside of the core flake (as you note.)

I don't _think_ I invented this pattern, but it's not one that you'll see in `nixpkgs`, since there's very little use of `apps` in non-flake repositories.

> Use of https://github.com/nix-systems/nix-systems

Of all the changes I'm the least attached to this one. I really wish that nixpkgs had a variable that contained `["x86-64_linux", "aarch64-linux", "x86_64-darwin", "aarch64-darwin"]`, but until nixos/nixpkgs#295083 is accepted and merged, no such list exists.

I'll revert this commit.

> 1000 instances

As I understand it, every use of `import nixpkgs` in a flake is subject to this essay. When you use it through `nixpkgs.legacyPackages.${system}`, you're letting nixpkgs be imported the way it wanted to be, rather than forcing an empty configuration on it, as flake.nix did before.

It's a pretty simple change that ostensibly has ecosystem effects, so I find myself happy to make it.

> systemClosure

Yeah, I think we can make it the genAttrs version.

**Comment by n8henrie** [comment]
Thanks for a thorough response and explanations!

I have no big issue with the callPackage stuff -- it is definitely the dominant paradigm, was mostly just questioning how useful the implementation would likely be in practice fort his project; I'm aware that there are plenty of theoretical benefits. If you have time, could you give me an example of how you'd "import it in my own overlay and fit it in"? For the moment it looks like you're [using this project](https://github.com/philiptaron/flock.nix/blob/4e6bb7d0ce689e6237728c00cc45afa4c89b32ab/nix.nix#L43) in a pretty straightforward way.

> As I understand it, every use of import nixpkgs in a flake is subject to this essay.

Good discussion here (including from the author of that essay): https://discourse.nixos.org/t/using-nixpkgs-legacypackages-system-vs-import/17462/19 . Even after reading that, I'm not sure that there's a benefit in this case at this point in time, as there is only 1 input to the flake, and the `import nixpkgs` only occurs once per system (for which `nixpkgs.legacyPackages.${system}` would have to be evaluated anyway). However, were there multiple instances of `import nixpkgs`, it looks like nix doesn't cache the result of this function call, so even identical `import nixpkgs {...}` calls will be separately evaluated (costing memory and time at evaluation time). Does that fit with your understanding? (Looks like I should probably update... a lot of flakes.) Thanks for bringing this to my attention!

**Comment by philiptaron** [comment]
## Use of `package.nix`
> For the moment it looks like you're [using this project](https://github.com/philiptaron/flock.nix/blob/4e6bb7d0ce689e6237728c00cc45afa4c89b32ab/nix.nix#L43) in a pretty straightforward way.

For sure, and one reason is that I wasn't able to integrate and use the flake, so I went with the copy that was in `nixpkgs`. 🙂 

My version is quite similar to your hypothesized version:

### `overlay.nix`
```nix
inputs: final: prev: 

{
  nix-bisect = final.myFunkyPython.pkgs.callPackage "${inputs.nix-bisect}/package.nix" { };
}
```

### `flake.nix`
```nix
{
   inputs = ...
   outputs = inputs: {
     ...
     overlays = {
       default = import ./overlays.nix inputs;
     };
     ...
   };
}
```

## 1000 instances of nixpkgs

Yeah, it's not about _this_ use, which as you point out is just once. It's about _every flake doing this_, which then means that users integrating more flakes have to re-instantiate `nixpkgs` **for every flake they depend on**...

If all the flakes take their nixpkgs dependencies from `inputs.nixpkgs.legacyPackages.${system}` instead, it's much better: `nixpkgs` gets instantiated once, and all the users pull from the same common source.

There's obviously more work to do here, and I expect the Nix team will do it; for now, my best practice is to switch `import nixpkgs { }` with `inputs.nixpkgs.legacyPackages.${system}` and call it a day, unless there's something extra that the flake needs. 

An example of something extra is CUDA support or non-free packages.

**Comment by timokau** [comment]
Thanks for the extensive review @n8henrie! I do not have the time to review this in detail right now, so I really appreciate it. Can you approve the PR once you are satisfied with the changes? I will review the PR afterwards, but it will likely take some time.

In case either or both of you are interested in maintaining nix-bisect, note that I actively encourage interested community members to fork this project. I will gladly link to a maintained fork in the readme. (as posted in #27)

**Comment by n8henrie** [comment]
Thanks for the additional context @timokau. I will keep in mind your invitation -- thank you. I don't think I have the permissions required to "approve" the PR, as I'm a `contributor` but not a `member`, but I'll gladly signal my approval with a comment!

@philiptaron -- no blockers from me, just a few other questions / thoughts for both of you:

- I'm most accustomed to using `default.nix` as a convenience for the `callPackage ./. {overrides}` pattern, in which case it seems like `default.nix` could be simplified to `{ python3 }: { python3.pkgs... }` (with overrides on the nixpkgs rev being bumped up a little higher). Unless overriding the rev is part of the bisect process used by nix-bisect itself (I'm not familiar enough with the codebase to know offhand). Not a big deal either way,.
- what would you think of following `nixpkgs-unstable` instead of `nixos-unstable`? My understanding is that the former has worse testing for NixOS but better testing for darwin (I'm primarily a darwin user); it's what I usually follow.

Other:
- this project seems like a great candidate for an integration test (which would also serve as documentation). I'll work on this eventually, but might be a while.
- @philiptaron I'm still interested in the "1000-instances-of-nixpkgs" idea -- do you have any suggestions on how to test / compare? Perhaps some kind of override that calls `builtins.trace` upon instantiating nixpkgs? Or maybe just tracing max memory usage when evaluating a derivation? Still interested in learning more about this.

**Comment by philiptaron** [comment]
## Why `package.nix` not `default.nix`?

[Because `pkgs/by-name` does it that way](https://github.com/NixOS/rfcs/blob/master/rfcs/0140-simple-package-paths.md), on the grounds that having a package (something that fits into [`callPackage`](https://nix.dev/tutorials/callpackage)) be named `package.nix` is clearer than naming it `default.nix`.

## Following `nixpkgs-unstable` not `nixos-unstable`

Frankly, at the rate of update, there's not much difference. This `flake.lock` is now two weeks old. I'll gladly follow whatever gets merged. 🙂 

## 1000 instances of nixpkgs

To observe:

Make 100 flakes, each exporting a single package that's (say) a trivial C library. They all do `import nixpkgs`. Then make a consuming flake that has all 100 as inputs. Building the final flake is going to cause Nix to consume `O(nixpkgs)` ✖️  `O(inputs)` memory.

This is the case even if every single one of those flakes is made to follow the exact same nixpkgs url. It's the act of `import nixpkgs` that causes the ballooning memory.

Then modify each of those flakes to pull their dependencies from `inputs.nixpkgs.legacyPackages.${system}`. Observe that the memory use from Nix is now `O(nixpkgs)`.

To be clear, this is an issue _for the consumer_ of the flake, not for _the flake itself_.

**Comment by n8henrie** [comment]
Thanks for the repro on the 1,000 instances issue -- I'll tinker with that!

I think I wasn't clear about `default.nix` -- my Q was whether the default.nix could be simplified, as most uses of the callPackage pattern pass in attributes of `pkgs` instead of `pkgs` itself.

**Comment by philiptaron** [comment]
The `default.nix` is there as an alternative to flakes. I'm loath to reshape it further.

**Comment by n8henrie** [comment]
No worries -- I've just never used the old-style commands, so I thought `default.nix` was only for the `callPackage` pattern, but obviously it's not.

https://nixos.org/manual/nix/unstable/command-ref/nix-build.html

> If no paths are specified, then nix-build will use default.nix in the current directory, if it exists.

I think this all looks good, read to merge.

**Comment by n8henrie** [comment]
@philiptaron -- my attempt at an MRE is showing  nearly twice as much memory consumption for `legacyPackages` as opposed to `import nixpkgs`, whereas I had expected something like 100x greater in the opposite direction.

I've started a thread at https://discourse.nixos.org/t/mre-for-1000-instances-of-nixpkgs-problem/43982 so as not to pollute this PR further. Hopefully you are the same Philip Taron I tagged there! Would love to sort through this more, thanks again for bringing this to my attention.

**Comment by philiptaron** [comment]
I am indeed. Let's continue the discussion there.

**Comment by philiptaron** [comment]
I'm planning on closing this in a week. I don't like unmerged PRs on my dashboard 🔢.

**Comment by n8henrie** [comment]
Fair enough. @timokau, FWIW I think this has some valuable changes and is worth merging.

**Review by timokau — APPROVED** [review]
Thank you (also for the reminder)!

**Pull request merged**
